src/extension/src/azure/azure-resource-group/resourceGroupModule.ts

```diff
@@ -0,0 +1,98 @@
+import { SubscriptionItem } from "../azure-auth/azureAuth";
+import { ResourceManager } from "../azure-arm/resourceManager";
+import { ResourceGroup } from "azure-arm-resource/lib/resource/models";
+import { ResourceManagementClient } from "azure-arm-resource/lib/resource/resourceManagementClient";
+import { ResourceGroupError, AuthorizationError } from "../../errors";
+import { CONSTANTS } from "../../constants";
+
+const VALIDATION_LIMIT = 3;
+
+export interface ResourceGroupSelection {
+    subscriptionItem: SubscriptionItem;
+    resourceGroupName: string;
+    location: string;
+}
+
+export class ResourceGroupDeploy {
+    private azureResourceClient: ResourceManagementClient | undefined;
+
+    private setAzureResourceClient (userSubscriptionItem: SubscriptionItem) {
+        if (
+            this.azureResourceClient === undefined ||
+            this.azureResourceClient.subscriptionId !== userSubscriptionItem.subscriptionId
+            ) {
+                this.azureResourceClient = new ResourceManager().getResourceManagementClient(userSubscriptionItem);
+            }
+    }
+
+    public async createResourceGroup(
+        resourceGroupSelection: ResourceGroupSelection
+    ): Promise<string | undefined> {
+        let userSubscriptionItem: SubscriptionItem = resourceGroupSelection.subscriptionItem;
+        let resourceGroupName: string = resourceGroupSelection.resourceGroupName;
+        let location: string = resourceGroupSelection.location;
+
+        let azureResourceClient: ResourceManagementClient = new ResourceManager().getResourceManagementClient(userSubscriptionItem);
+
+        let options: ResourceGroup = {
+            name: resourceGroupName,
+            location: location,
+        };
+
+        let result = await azureResourceClient.resourceGroups.createOrUpdate(
+            resourceGroupName, 
+            options
+            ); 
+        return result.name;
+    }
+
+    public async generateValidResourceGroupName(
+        name: string, 
+        userSubscriptionItem: SubscriptionItem
+    ): Promise<string> {
+        let tries = 0;
+        this.setAzureResourceClient(userSubscriptionItem);
+        let generatedName = this.generateResourceGroupName(name);
+        let isValid = await this.validateResourceGroupName(generatedName);
+
+        while (tries < VALIDATION_LIMIT && !isValid) {
+            generatedName = this.generateResourceGroupName(name);
+            isValid = await this.validateResourceGroupName(generatedName);
+            tries++;
```

Comment:
What does it mean if tries exceeds validation limit? Does it mean Azure is down or network connectivity problem, or do you think it does not matter to the user?

Reply:
It is very unlikely that tries will exceed the validation limit because a new name (with a new timestamp) will be generated before it will be validated again. 

Just to address your question, I tried disconnecting my computer, and I was not able to add any Azure resources, which means that this method will never be trigerred if the user has connectivity problem.

So, if there is ever a case that tries exceed the validation limit, it is more likely that Azure is down. 

Please let me know if there is any other edge cases I should consider.

Reply:
There is another edge case where the user signs out of their azure account using the command pallet command rather than through the wizard after adding azure. But this is a known issue which can't be fixed yet, or at least can't be fixed nicely.  There is a PR up for the API solution that would help us, but the vscode team is still doing tests on it to see if there are any performance implications.  (I'm following this issue really closely).